images/handlers.go

```diff
@@ -179,14 +181,102 @@ func Dispatch(ctx context.Context, handler Handler, limiter *semaphore.Weighted,
 			if len(children) > 0 {
 				return Dispatch(ctx2, handler, limiter, children...)
 			}
-
 			return nil
 		})
 	}
+	return eg.Wait()
+}
+
+// DispatchWithOpts runs the provided handler for content specified by the
+// descriptors with concurrent and duplication suppressor control.
+//
+// If the handler decodes subresources, they will be visited, as well. A handler
+// can return 'ErrSkipDesc` to signal to the dispatcher to not traverse any
+// children. If any handler returns an error, the dispatch session will be
+// canceled.
+func DispatchWithOpts(ctx context.Context, handler Handler, opts []DispatchOpt, descs ...ocispec.Descriptor) error {
```

Comment:
Instead of building support for duplication suppression directly into the dispatch code, we could instead leave it up to the caller to pass a handler wrapper that implements the suppression themselves. So rather than adding this function and `WithDuplicationSuppressor`, we could instead have CRI wrap [`appendInfoHandlerWrapper`](https://github.com/containerd/containerd/blob/main/pkg/cri/server/image_pull.go#L531) in another wrapper which uses `kmutex` to implement suppression.

This would give the client more control, and keep the "core" client code cleaner.

Reply:
Yes! Thanks for the comment. I added `do-not-merge` label because I found that the HandlerWrapper can handle this. 
https://github.com/fuweid/containerd/pull/2.